File: src/test/java/wile/redstonepen/libmc/InventoryRangeTest.java
package wile.redstonepen.libmc;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;

import net.minecraft.world.SimpleContainer;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import org.junit.jupiter.api.Test;

class InventoryRangeTest
{
  private static SimpleContainer container(int size)
  { return new SimpleContainer(size); }

  private static SimpleContainer filled(int size, int redstone)
  {
    final SimpleContainer c = new SimpleContainer(size);
    for(int i=0; i<size; ++i) c.setItem(i, new ItemStack(Items.REDSTONE, redstone));
    return c;
  }

  // ---- construction / accessors ---------------------------------------------------------------

  @Test
  void constructorClampsOffsetAndSizeToContainerBounds()
  {
    final SimpleContainer c = container(9);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c, 100, 100, 1);
    assertTrue(r.size() <= 9);
    assertTrue(r.offset() >= 0);
    assertEquals(c, r.inventory());
  }

  @Test
  void singleArgConstructorWrapsEntireContainer()
  {
    final SimpleContainer c = container(7);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    assertEquals(7, r.size());
    assertEquals(0, r.offset());
    assertEquals(7, r.getContainerSize());
  }

  @Test
  void twoArgConstructorDefaultsRowsToOne()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(9), 0, 9);
    assertEquals(9, r.size());
  }

  @Test
  void getAndSetUseTheRangeOffset()
  {
    final SimpleContainer c = container(9);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c, 3, 4, 1);
    r.set(0, new ItemStack(Items.REDSTONE, 2));
    assertEquals(Items.REDSTONE, c.getItem(3).getItem());
    assertEquals(2, r.get(0).getCount());
  }

  @Test
  void maxStackSizeIsBoundedAndAtLeastOne()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(4));
    r.setMaxStackSize(0);
    assertTrue(r.getMaxStackSize() >= 1);
    r.setMaxStackSize(8);
    assertEquals(8, r.getMaxStackSize());
  }

  @Test
  void validatorCanRejectItems()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(4));
    r.setValidator((slot, stack) -> stack.is(Items.REDSTONE));
    assertNotNull(r.getValidator());
    assertTrue(r.canPlaceItem(0, new ItemStack(Items.REDSTONE)));
    assertFalse(r.canPlaceItem(0, new ItemStack(Items.COAL)));
  }

  // ---- Container delegates --------------------------------------------------------------------

  @Test
  void clearContentEmptiesEverySlotInRange()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(filled(4, 2));
    r.clearContent();
    assertTrue(r.isEmpty());
  }

  @Test
  void removeItemNoUpdateRemovesAndReturnsStack()
  {
    final SimpleContainer c = filled(2, 4);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack out = r.removeItemNoUpdate(0);
    assertEquals(4, out.getCount());
    assertTrue(r.getItem(0).isEmpty());
  }

  @Test
  void removeItemSplitsStack()
  {
    final SimpleContainer c = filled(2, 5);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack out = r.removeItem(0, 3);
    assertEquals(3, out.getCount());
    assertEquals(2, r.getItem(0).getCount());
  }

  @Test
  void stillValidDelegatesToInventory()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(1));
    assertTrue(r.stillValid(null));
  }

  @Test
  void startAndStopOpenAreNoOps()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(1));
    r.startOpen(null);
    r.stopOpen(null);
    r.setChanged();
  }

  // ---- iterate / contains / indexOf / find / collect / stream / iterator ---------------------

  @Test
  void iterateStopsWhenPredicateMatches()
  {
    final SimpleContainer c = container(3);
    c.setItem(1, new ItemStack(Items.COAL));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final List<Integer> seen = new ArrayList<>();
    final boolean matched = r.iterate((i, s) -> { seen.add(i); return s.is(Items.COAL); });
    assertTrue(matched);
    assertEquals(List.of(0, 1), seen);
  }

  @Test
  void iterateReturnsFalseWhenNothingMatches()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(filled(3, 1));
    assertFalse(r.iterate((i, s) -> false));
  }

  @Test
  void containsAndIndexOfFindMatchingStack()
  {
    final SimpleContainer c = container(3);
    c.setItem(2, new ItemStack(Items.REDSTONE, 1));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    assertTrue(r.contains(new ItemStack(Items.REDSTONE, 1)));
    assertEquals(2, r.indexOf(new ItemStack(Items.REDSTONE, 1)));
    assertEquals(-1, r.indexOf(new ItemStack(Items.COAL, 1)));
  }

  @Test
  void findReturnsFirstPresentResult()
  {
    final SimpleContainer c = container(3);
    c.setItem(0, new ItemStack(Items.COAL));
    c.setItem(2, new ItemStack(Items.REDSTONE));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final Optional<Integer> idx = r.find((i, s) -> s.is(Items.REDSTONE) ? Optional.of(i) : Optional.empty());
    assertEquals(Optional.of(2), idx);
  }

  @Test
  void findReturnsEmptyWhenNothingMatches()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(2));
    assertTrue(r.find((i, s) -> Optional.<Integer>empty()).isEmpty());
  }

  @Test
  void collectGathersAllPresentResults()
  {
    final SimpleContainer c = container(4);
    c.setItem(0, new ItemStack(Items.REDSTONE));
    c.setItem(2, new ItemStack(Items.REDSTONE));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final List<Integer> hits = r.collect((i, s) -> s.is(Items.REDSTONE) ? Optional.of(i) : Optional.empty());
    assertEquals(List.of(0, 2), hits);
  }

  @Test
  void streamYieldsAllSlotsIncludingEmpty()
  {
    final SimpleContainer c = container(3);
    c.setItem(1, new ItemStack(Items.REDSTONE, 4));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    assertEquals(3, r.stream().count());
  }

  @Test
  void iteratorWalksRangeAndThrowsAtEnd()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(filled(2, 1));
    final Iterator<ItemStack> it = r.iterator();
    assertTrue(it.hasNext()); it.next();
    assertTrue(it.hasNext()); it.next();
    assertFalse(it.hasNext());
    org.junit.jupiter.api.Assertions.assertThrows(NoSuchElementException.class, it::next);
  }

  // ---- match counts ---------------------------------------------------------------------------

  @Test
  void stackMatchCountAndTotalCount()
  {
    final SimpleContainer c = container(4);
    c.setItem(0, new ItemStack(Items.REDSTONE, 3));
    c.setItem(1, new ItemStack(Items.COAL, 5));
    c.setItem(3, new ItemStack(Items.REDSTONE, 7));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    assertEquals(2, r.stackMatchCount(new ItemStack(Items.REDSTONE, 3)));
    assertEquals(10, r.totalMatchingItemCount(new ItemStack(Items.REDSTONE, 3)));
  }

  // ---- insert ---------------------------------------------------------------------------------

  @Test
  void insertEmptyStackReturnsEmpty()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(3));
    assertTrue(r.insert(ItemStack.EMPTY).isEmpty());
  }

  @Test
  void insertFillsExistingMatchingStackBeforeAnyEmptySlot()
  {
    final SimpleContainer c = container(3);
    c.setItem(2, new ItemStack(Items.REDSTONE, 10));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack remaining = r.insert(new ItemStack(Items.REDSTONE, 5));
    assertTrue(remaining.isEmpty());
    assertEquals(15, c.getItem(2).getCount());
    assertTrue(c.getItem(0).isEmpty());
  }

  @Test
  void insertReturnsOverflowWhenNoCapacity()
  {
    final SimpleContainer c = container(1);
    c.setItem(0, new ItemStack(Items.REDSTONE, 64));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack remaining = r.insert(new ItemStack(Items.REDSTONE, 5));
    assertEquals(5, remaining.getCount());
  }

  @Test
  void insertOnlyFillupSkipsEmptySlots()
  {
    final SimpleContainer c = container(3);
    c.setItem(0, new ItemStack(Items.REDSTONE, 60));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack remaining = r.insert(new ItemStack(Items.REDSTONE, 10), true, 0, false, false);
    assertEquals(64, c.getItem(0).getCount());
    assertEquals(6, remaining.getCount());
    assertTrue(c.getItem(1).isEmpty());
  }

  @Test
  void insertUsesEmptySlotWhenNoMatchExists()
  {
    final SimpleContainer c = container(3);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack remaining = r.insert(new ItemStack(Items.REDSTONE, 4));
    assertTrue(remaining.isEmpty());
    assertEquals(Items.REDSTONE, c.getItem(0).getItem());
    assertEquals(4, c.getItem(0).getCount());
  }

  @Test
  void insertReverseFillsFromTheEnd()
  {
    final SimpleContainer c = container(3);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    r.insert(new ItemStack(Items.REDSTONE, 4), false, 0, true, true);
    assertEquals(Items.REDSTONE, c.getItem(2).getItem());
  }

  @Test
  void insertSimulateDoesNotMutate()
  {
    final SimpleContainer c = container(2);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack remaining = r.insert(new ItemStack(Items.REDSTONE, 4), true);
    assertTrue(remaining.isEmpty());
    assertTrue(c.getItem(0).isEmpty());
  }

  @Test
  void insertSimulateReportsLeftoverWhenFull()
  {
    final SimpleContainer c = container(1);
    c.setItem(0, new ItemStack(Items.REDSTONE, 64));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack remaining = r.insert(new ItemStack(Items.REDSTONE, 4), true);
    assertEquals(4, remaining.getCount());
  }

  @Test
  void insertAtIndexFillsTargetSlotAndReturnsRemainder()
  {
    final SimpleContainer c = container(2);
    c.setItem(0, new ItemStack(Items.REDSTONE, 60));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack remaining = r.insert(0, new ItemStack(Items.REDSTONE, 10));
    assertEquals(64, c.getItem(0).getCount());
    assertEquals(6, remaining.getCount());
  }

  @Test
  void insertAtIndexIntoEmptyPlacesFullStack()
  {
    final SimpleContainer c = container(2);
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack remaining = r.insert(0, new ItemStack(Items.REDSTONE, 10));
    assertTrue(remaining.isEmpty());
    assertEquals(10, c.getItem(0).getCount());
  }

  @Test
  void insertAtIndexIntoMismatchedSlotReturnsInputUnchanged()
  {
    final SimpleContainer c = container(2);
    c.setItem(0, new ItemStack(Items.COAL, 4));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack input = new ItemStack(Items.REDSTONE, 3);
    final ItemStack remaining = r.insert(0, input);
    assertEquals(3, remaining.getCount());
    assertEquals(Items.REDSTONE, remaining.getItem());
  }

  @Test
  void insertAtIndexEmptyInputReturnsEmpty()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(1));
    assertTrue(r.insert(0, ItemStack.EMPTY).isEmpty());
  }

  // ---- extract --------------------------------------------------------------------------------

  @Test
  void extractAmountFromEmptyRangeReturnsEmpty()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(container(3));
    assertTrue(r.extract(5).isEmpty());
  }

  @Test
  void extractTakesFromFirstNonEmptyStack()
  {
    final SimpleContainer c = container(3);
    c.setItem(1, new ItemStack(Items.REDSTONE, 10));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack out = r.extract(4);
    assertEquals(4, out.getCount());
    assertEquals(6, c.getItem(1).getCount());
  }

  @Test
  void extractAggregatesAcrossIdenticalStacks()
  {
    final SimpleContainer c = container(3);
    c.setItem(0, new ItemStack(Items.REDSTONE, 3));
    c.setItem(1, new ItemStack(Items.COAL, 8));
    c.setItem(2, new ItemStack(Items.REDSTONE, 2));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack out = r.extract(5);
    assertEquals(Items.REDSTONE, out.getItem());
    assertEquals(5, out.getCount());
    assertTrue(c.getItem(0).isEmpty());
    assertTrue(c.getItem(2).isEmpty());
    assertEquals(8, c.getItem(1).getCount()); // untouched
  }

  @Test
  void extractSimulateDoesNotMutate()
  {
    final SimpleContainer c = container(2);
    c.setItem(0, new ItemStack(Items.REDSTONE, 5));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack out = r.extract(2, false, true);
    assertEquals(2, out.getCount());
    assertEquals(5, c.getItem(0).getCount());
  }

  @Test
  void extractByStackReturnsEmptyForEmptyRequest()
  {
    final Inventories.InventoryRange r = new Inventories.InventoryRange(filled(2, 4));
    assertTrue(r.extract(ItemStack.EMPTY).isEmpty());
  }

  @Test
  void extractByStackPullsRequestedAmount()
  {
    final SimpleContainer c = container(3);
    c.setItem(0, new ItemStack(Items.REDSTONE, 4));
    c.setItem(2, new ItemStack(Items.REDSTONE, 3));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack req = new ItemStack(Items.REDSTONE, 5);
    final ItemStack out = r.extract(req);
    assertEquals(5, out.getCount());
  }

  @Test
  void extractByStackSimulateReturnsAvailableAmountAndDoesNotMutate()
  {
    final SimpleContainer c = container(3);
    c.setItem(0, new ItemStack(Items.REDSTONE, 4));
    c.setItem(2, new ItemStack(Items.REDSTONE, 3));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    final ItemStack out = r.extract(new ItemStack(Items.REDSTONE, 50), true);
    assertEquals(7, out.getCount());
    assertEquals(4, c.getItem(0).getCount());
    assertEquals(3, c.getItem(2).getCount());
  }

  @Test
  void extractByStackReturnsEmptyWhenNoMatch()
  {
    final SimpleContainer c = container(2);
    c.setItem(0, new ItemStack(Items.COAL, 4));
    final Inventories.InventoryRange r = new Inventories.InventoryRange(c);
    assertTrue(r.extract(new ItemStack(Items.REDSTONE, 1)).isEmpty());
  }

  // ---- move -----------------------------------------------------------------------------------

  @Test
  void moveSlotMovesIntoEmptyTarget()
  {
    final SimpleContainer src = container(2);
    src.setItem(0, new ItemStack(Items.REDSTONE, 5));
    final SimpleContainer dst = container(2);
    final Inventories.InventoryRange srcR = new Inventories.InventoryRange(src);
    final Inventories.InventoryRange dstR = new Inventories.InventoryRange(dst);
    assertTrue(srcR.move(0, dstR));
    assertTrue(src.getItem(0).isEmpty());
    assertEquals(5, dst.getItem(0).getCount());
  }

  @Test
  void moveSlotEmptySourceReturnsFalse()
  {
    final Inventories.InventoryRange src = new Inventories.InventoryRange(container(2));
    final Inventories.InventoryRange dst = new Inventories.InventoryRange(container(2));
    assertFalse(src.move(0, dst));
  }

  @Test
  void moveAllIdenticalDrainsSourceWhenTargetHasRoom()
  {
    final SimpleContainer src = container(3);
    src.setItem(0, new ItemStack(Items.REDSTONE, 30));
    src.setItem(1, new ItemStack(Items.REDSTONE, 20));
    src.setItem(2, new ItemStack(Items.REDSTONE, 10));
    final SimpleContainer dst = container(2);
    final Inventories.InventoryRange srcR = new Inventories.InventoryRange(src);
    final Inventories.InventoryRange dstR = new Inventories.InventoryRange(dst);
    assertTrue(srcR.move(0, dstR, true, false, false, true));
  }

  @Test
  void moveRangeMovesEverythingPossible()
  {
    final SimpleContainer src = container(3);
    src.setItem(0, new ItemStack(Items.REDSTONE, 5));
    src.setItem(2, new ItemStack(Items.COAL, 3));
    final SimpleContainer dst = container(3);
    final Inventories.InventoryRange srcR = new Inventories.InventoryRange(src);
    final Inventories.InventoryRange dstR = new Inventories.InventoryRange(dst);
    assertTrue(srcR.move(dstR));
    assertTrue(src.getItem(0).isEmpty());
    assertTrue(src.getItem(2).isEmpty());
  }

  @Test
  void moveRangeReturnsFalseWhenSourceEmpty()
  {
    final Inventories.InventoryRange srcR = new Inventories.InventoryRange(container(3));
    final Inventories.InventoryRange dstR = new Inventories.InventoryRange(container(3));
    assertFalse(srcR.move(dstR));
  }

  @Test
  void moveRangeFillupVariantsCompile()
  {
    final SimpleContainer src = container(2);
    src.setItem(0, new ItemStack(Items.REDSTONE, 5));
    final SimpleContainer dst = container(2);
    final Inventories.InventoryRange srcR = new Inventories.InventoryRange(src);
    final Inventories.InventoryRange dstR = new Inventories.InventoryRange(dst);
    srcR.move(dstR, true);
  }

  // ---- module-level helpers -------------------------------------------------------------------

  @Test
  void multiRangeInsertConsumesUntilEmpty()
  {
    final SimpleContainer a = container(1);
    a.setItem(0, new ItemStack(Items.REDSTONE, 60));
    final SimpleContainer b = container(2);
    final Inventories.InventoryRange[] ranges = {
      new Inventories.InventoryRange(a),
      new Inventories.InventoryRange(b),
    };
    final ItemStack remaining = Inventories.insert(ranges, new ItemStack(Items.REDSTONE, 10));
    assertTrue(remaining.isEmpty());
  }
}

File: common/src/test/kotlin/wile/redstonepen/util/InventoryRangeTest.kt
package wile.redstonepen.util

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import java.util.NoSuchElementException
import java.util.Optional
import net.minecraft.world.SimpleContainer
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items

class InventoryRangeTest :
    DescribeSpec({
        fun container(size: Int) = SimpleContainer(size)

        fun filled(size: Int, redstone: Int) =
            SimpleContainer(size).also { c ->
                for (i in 0 until size) c.setItem(i, ItemStack(Items.REDSTONE, redstone))
            }

        describe("construction") {
            it("clamps offset and size to container bounds") {
                val c = container(9)
                val r = Inventories.InventoryRange(c, 100, 100, 1)
                (r.size() <= 9) shouldBe true
                (r.offset() >= 0) shouldBe true
                r.inventory() shouldBe c
            }

            it("single arg constructor wraps entire container") {
                val c = container(7)
                val r = Inventories.InventoryRange(c)
                r.size() shouldBe 7
                r.offset() shouldBe 0
                r.containerSize shouldBe 7
            }

            it("two arg constructor defaults rows to one") {
                Inventories.InventoryRange(container(9), 0, 9).size() shouldBe 9
            }

            it("get and set use the range offset") {
                val c = container(9)
                val r = Inventories.InventoryRange(c, 3, 4, 1)
                r.set(0, ItemStack(Items.REDSTONE, 2))
                c.getItem(3).item shouldBe Items.REDSTONE
                r.get(0).count shouldBe 2
            }

            it("maxStackSize is bounded and at least one") {
                val r = Inventories.InventoryRange(container(4))
                r.setMaxStackSize(0)
                (r.maxStackSize >= 1) shouldBe true
                r.setMaxStackSize(8)
                r.maxStackSize shouldBe 8
            }

            it("validator can reject items") {
                val r = Inventories.InventoryRange(container(4))
                r.setValidator { _, stack -> stack.`is`(Items.REDSTONE) }
                r.getValidator() shouldNotBe null
                r.canPlaceItem(0, ItemStack(Items.REDSTONE)) shouldBe true
                r.canPlaceItem(0, ItemStack(Items.COAL)) shouldBe false
            }

            it("multi-range insert consumes until empty") {
                val a = container(1)
                a.setItem(0, ItemStack(Items.REDSTONE, 60))
                val b = container(2)
                val ranges = arrayOf(Inventories.InventoryRange(a), Inventories.InventoryRange(b))
                Inventories.insert(ranges, ItemStack(Items.REDSTONE, 10)).isEmpty shouldBe true
            }
        }

        describe("container delegates") {
            it("clearContent empties every slot in range") {
                val r = Inventories.InventoryRange(filled(4, 2))
                r.clearContent()
                r.isEmpty shouldBe true
            }

            it("removeItemNoUpdate removes and returns stack") {
                val c = filled(2, 4)
                val r = Inventories.InventoryRange(c)
                val out = r.removeItemNoUpdate(0)
                out.count shouldBe 4
                r.getItem(0).isEmpty shouldBe true
            }

            it("removeItem splits stack") {
                val c = filled(2, 5)
                val r = Inventories.InventoryRange(c)
                val out = r.removeItem(0, 3)
                out.count shouldBe 3
                r.getItem(0).count shouldBe 2
            }

            it("stillValid delegates to inventory") {
                Inventories.InventoryRange(container(1)).stillValid(null) shouldBe true
            }

            it("startOpen stopOpen setChanged are no-ops") {
                val r = Inventories.InventoryRange(container(1))
                r.startOpen(null)
                r.stopOpen(null)
                r.setChanged()
            }
        }

        describe("iteration") {
            it("iterate stops when predicate matches") {
                val c = container(3)
                c.setItem(1, ItemStack(Items.COAL))
                val r = Inventories.InventoryRange(c)
                val seen = mutableListOf<Int>()
                val matched = r.iterate { i, s ->
                    seen.add(i)
                    s.`is`(Items.COAL)
                }
                matched shouldBe true
                seen shouldBe listOf(0, 1)
            }

            it("iterate returns false when nothing matches") {
                Inventories.InventoryRange(filled(3, 1)).iterate { _, _ -> false } shouldBe false
            }

            it("contains and indexOf find matching stack") {
                val c = container(3)
                c.setItem(2, ItemStack(Items.REDSTONE, 1))
                val r = Inventories.InventoryRange(c)
                r.contains(ItemStack(Items.REDSTONE, 1)) shouldBe true
                r.indexOf(ItemStack(Items.REDSTONE, 1)) shouldBe 2
                r.indexOf(ItemStack(Items.COAL, 1)) shouldBe -1
            }

            it("find returns first present result") {
                val c = container(3)
                c.setItem(0, ItemStack(Items.COAL))
                c.setItem(2, ItemStack(Items.REDSTONE))
                val r = Inventories.InventoryRange(c)
                val idx = r.find { i, s ->
                    if (s.`is`(Items.REDSTONE)) Optional.of(i) else Optional.empty()
                }
                idx shouldBe Optional.of(2)
            }

            it("find returns empty when nothing matches") {
                Inventories.InventoryRange(container(2))
                    .find { _, _ -> Optional.empty<Int>() }
                    .isEmpty shouldBe true
            }

            it("collect gathers all present results") {
                val c = container(4)
                c.setItem(0, ItemStack(Items.REDSTONE))
                c.setItem(2, ItemStack(Items.REDSTONE))
                val r = Inventories.InventoryRange(c)
                val hits = r.collect { i, s ->
                    if (s.`is`(Items.REDSTONE)) Optional.of(i) else Optional.empty()
                }
                hits shouldBe listOf(0, 2)
            }

            it("stream yields all slots including empty") {
                val c = container(3)
                c.setItem(1, ItemStack(Items.REDSTONE, 4))
                Inventories.InventoryRange(c).stream().count() shouldBe 3
            }

            it("iterator walks range and throws at end") {
                val r = Inventories.InventoryRange(filled(2, 1))
                val it = r.iterator()
                it.hasNext() shouldBe true
                it.next()
                it.hasNext() shouldBe true
                it.next()
                it.hasNext() shouldBe false
                shouldThrow<NoSuchElementException> { it.next() }
            }
        }

        describe("match counts") {
            it("stackMatchCount and totalMatchingItemCount") {
                val c = container(4)
                c.setItem(0, ItemStack(Items.REDSTONE, 3))
                c.setItem(1, ItemStack(Items.COAL, 5))
                c.setItem(3, ItemStack(Items.REDSTONE, 7))
                val r = Inventories.InventoryRange(c)
                r.stackMatchCount(ItemStack(Items.REDSTONE, 3)) shouldBe 2
                r.totalMatchingItemCount(ItemStack(Items.REDSTONE, 3)) shouldBe 10
            }
        }

        describe("insert") {
            it("empty stack returns empty") {
                Inventories.InventoryRange(container(3)).insert(ItemStack.EMPTY).isEmpty shouldBe
                    true
            }

            it("fills existing matching stack before any empty slot") {
                val c = container(3)
                c.setItem(2, ItemStack(Items.REDSTONE, 10))
                val r = Inventories.InventoryRange(c)
                r.insert(ItemStack(Items.REDSTONE, 5)).isEmpty shouldBe true
                c.getItem(2).count shouldBe 15
                c.getItem(0).isEmpty shouldBe true
            }

            it("returns overflow when no capacity") {
                val c = container(1)
                c.setItem(0, ItemStack(Items.REDSTONE, 64))
                Inventories.InventoryRange(c).insert(ItemStack(Items.REDSTONE, 5)).count shouldBe 5
            }

            it("fillup-only skips empty slots") {
                val c = container(3)
                c.setItem(0, ItemStack(Items.REDSTONE, 60))
                val r = Inventories.InventoryRange(c)
                val remaining = r.insert(ItemStack(Items.REDSTONE, 10), true, 0, false, false)
                c.getItem(0).count shouldBe 64
                remaining.count shouldBe 6
                c.getItem(1).isEmpty shouldBe true
            }

            it("uses empty slot when no match exists") {
                val c = container(3)
                val r = Inventories.InventoryRange(c)
                r.insert(ItemStack(Items.REDSTONE, 4)).isEmpty shouldBe true
                c.getItem(0).item shouldBe Items.REDSTONE
                c.getItem(0).count shouldBe 4
            }

            it("reverse fills from the end") {
                val c = container(3)
                val r = Inventories.InventoryRange(c)
                r.insert(ItemStack(Items.REDSTONE, 4), false, 0, true, true)
                c.getItem(2).item shouldBe Items.REDSTONE
            }

            it("simulate does not mutate") {
                val c = container(2)
                val r = Inventories.InventoryRange(c)
                r.insert(ItemStack(Items.REDSTONE, 4), true).isEmpty shouldBe true
                c.getItem(0).isEmpty shouldBe true
            }

            it("simulate reports leftover when full") {
                val c = container(1)
                c.setItem(0, ItemStack(Items.REDSTONE, 64))
                Inventories.InventoryRange(c)
                    .insert(ItemStack(Items.REDSTONE, 4), true)
                    .count shouldBe 4
            }

            it("at index fills target slot and returns remainder") {
                val c = container(2)
                c.setItem(0, ItemStack(Items.REDSTONE, 60))
                val r = Inventories.InventoryRange(c)
                val remaining = r.insert(0, ItemStack(Items.REDSTONE, 10))
                c.getItem(0).count shouldBe 64
                remaining.count shouldBe 6
            }

            it("at index into empty places full stack") {
                val c = container(2)
                val r = Inventories.InventoryRange(c)
                r.insert(0, ItemStack(Items.REDSTONE, 10)).isEmpty shouldBe true
                c.getItem(0).count shouldBe 10
            }

            it("at index into mismatched slot returns input unchanged") {
                val c = container(2)
                c.setItem(0, ItemStack(Items.COAL, 4))
                val r = Inventories.InventoryRange(c)
                val remaining = r.insert(0, ItemStack(Items.REDSTONE, 3))
                remaining.count shouldBe 3
                remaining.item shouldBe Items.REDSTONE
            }

            it("at index empty input returns empty") {
                Inventories.InventoryRange(container(1)).insert(0, ItemStack.EMPTY).isEmpty shouldBe
                    true
            }
        }

        describe("extract") {
            it("from empty range returns empty") {
                Inventories.InventoryRange(container(3)).extract(5).isEmpty shouldBe true
            }

            it("takes from first non-empty stack") {
                val c = container(3)
                c.setItem(1, ItemStack(Items.REDSTONE, 10))
                val r = Inventories.InventoryRange(c)
                val out = r.extract(4)
                out.count shouldBe 4
                c.getItem(1).count shouldBe 6
            }

            it("aggregates across identical stacks") {
                val c = container(3)
                c.setItem(0, ItemStack(Items.REDSTONE, 3))
                c.setItem(1, ItemStack(Items.COAL, 8))
                c.setItem(2, ItemStack(Items.REDSTONE, 2))
                val r = Inventories.InventoryRange(c)
                val out = r.extract(5)
                out.item shouldBe Items.REDSTONE
                out.count shouldBe 5
                c.getItem(0).isEmpty shouldBe true
                c.getItem(2).isEmpty shouldBe true
                c.getItem(1).count shouldBe 8
            }

            it("simulate does not mutate") {
                val c = container(2)
                c.setItem(0, ItemStack(Items.REDSTONE, 5))
                val r = Inventories.InventoryRange(c)
                r.extract(2, false, true).count shouldBe 2
                c.getItem(0).count shouldBe 5
            }

            it("by stack returns empty for empty request") {
                Inventories.InventoryRange(filled(2, 4)).extract(ItemStack.EMPTY).isEmpty shouldBe
                    true
            }

            it("by stack pulls requested amount") {
                val c = container(3)
                c.setItem(0, ItemStack(Items.REDSTONE, 4))
                c.setItem(2, ItemStack(Items.REDSTONE, 3))
                Inventories.InventoryRange(c).extract(ItemStack(Items.REDSTONE, 5)).count shouldBe 5
            }

            it("by stack simulate returns available amount and does not mutate") {
                val c = container(3)
                c.setItem(0, ItemStack(Items.REDSTONE, 4))
                c.setItem(2, ItemStack(Items.REDSTONE, 3))
                val r = Inventories.InventoryRange(c)
                val out = r.extract(ItemStack(Items.REDSTONE, 50), true)
                out.count shouldBe 7
                c.getItem(0).count shouldBe 4
                c.getItem(2).count shouldBe 3
            }

            it("by stack returns empty when no match") {
                val c = container(2)
                c.setItem(0, ItemStack(Items.COAL, 4))
                Inventories.InventoryRange(c).extract(ItemStack(Items.REDSTONE, 1)).isEmpty shouldBe
                    true
            }
        }

        describe("move") {
            it("slot moves into empty target") {
                val src = container(2)
                src.setItem(0, ItemStack(Items.REDSTONE, 5))
                val dst = container(2)
                val srcR = Inventories.InventoryRange(src)
                val dstR = Inventories.InventoryRange(dst)
                srcR.move(0, dstR) shouldBe true
                src.getItem(0).isEmpty shouldBe true
                dst.getItem(0).count shouldBe 5
            }

            it("slot with empty source returns false") {
                Inventories.InventoryRange(container(2))
                    .move(0, Inventories.InventoryRange(container(2))) shouldBe false
            }

            it("moveAll identical drains source when target has room") {
                val src = container(3)
                src.setItem(0, ItemStack(Items.REDSTONE, 30))
                src.setItem(1, ItemStack(Items.REDSTONE, 20))
                src.setItem(2, ItemStack(Items.REDSTONE, 10))
                val srcR = Inventories.InventoryRange(src)
                val dstR = Inventories.InventoryRange(container(2))
                srcR.move(0, dstR, true, false, false, true) shouldBe true
            }

            it("range moves everything possible") {
                val src = container(3)
                src.setItem(0, ItemStack(Items.REDSTONE, 5))
                src.setItem(2, ItemStack(Items.COAL, 3))
                val dst = container(3)
                val srcR = Inventories.InventoryRange(src)
                val dstR = Inventories.InventoryRange(dst)
                srcR.move(dstR) shouldBe true
                src.getItem(0).isEmpty shouldBe true
                src.getItem(2).isEmpty shouldBe true
            }

            it("range returns false when source empty") {
                Inventories.InventoryRange(container(3))
                    .move(Inventories.InventoryRange(container(3))) shouldBe false
            }

            it("fillup variant compiles and runs") {
                val src = container(2)
                src.setItem(0, ItemStack(Items.REDSTONE, 5))
                val srcR = Inventories.InventoryRange(src)
                val dstR = Inventories.InventoryRange(container(2))
                srcR.move(dstR, true)
            }
        }
    })
